### PR TITLE
Deprecate `assert_display_snapshot`, refactor macros

### PR DESCRIPTION
**Pull request by max-sixty**
This PR:
- Deprecates `assert_display_snapshot` in favor of `assert_snapshot`, as discussed in #379
- Makes a sizable change in how macros are constructed:
  - It makes macros consistent — previously `debug` & `display` had different forms than the others, not taking a `debug_expr`.
  - The non-redaction macros are a simple wrapper of `_assert_snapshot_base`, which takes a closure `transform` of how to transform the input to a string.
  - The redaction macros are a simple wrapper of `_assert_serialized_snapshot`, which takes an arg `format` (similar to before), and uses that to construct a transform.
  - As a result, the boilerplate for each macros is removed.

~It's not quite complete, but wanted to post before spending more time on it.~ Now complete

Previous todos:
- [x] We get a couple of test failures; I need to look into this; we seem to be generating an additional `---` in some yaml snapshots. (possibly something is being converted to string twice?)
- [x] Update docs; previously the debug_expr wasn't well documented, and now that the source is more opaque this is more important

### CODE REVIEW

**Comment by max-sixty** [comment]
This is updated, with all the tests passing, and better comments around the macros.

One thing that would make the code even simpler — serializing values the same for `File` and `Inline`. I _think_ the only difference is that YAML snapshots lose the leading `---` in files — is that right? Any reason not to remove for both files & inline?

**Comment by max-sixty** [comment]
Updated the docs at https://github.com/mitsuhiko/insta-website/pull/17

**Comment by mitsuhiko** [comment]
I will try to get this reviewed this week.

File: src/macros.rs [hunk 1]
```diff
@@ -375,106 +297,90 @@ macro_rules! _prepare_snapshot_for_redaction {
...
+            $crate::_macro_support::ReferenceValue::Inline($snapshot),
+            $($arg),*
+        )
+    };
+    // If there's no debug_expr, use the stringified value, call self.
+    (transform=$transform:expr, $name:expr, $value:expr) => {
+        $crate::_assert_snapshot_base!(transform = $transform, $name, $value, stringify!($value))
+    };
+    // If there's no name (and necessarily no debug expr), auto generate the
+    // name and debug expr, call self.
```
Comment (max-sixty) [inline src/macros.rs:331]:
```suggestion
    // name, call self.
```

**Comment by max-sixty** [comment]
@mitsuhiko the gentlest of pings! Also will take feedback if you think it's not so great :)

**Comment by mitsuhiko** [comment]
I am in favor of this. I will try to get it merged and released ahead of backwards incompatible changes.

**Comment by max-sixty** [comment]
> I am in favor of this. I will try to get it merged and released ahead of backwards incompatible changes.

Thanks!

**Comment by kennethloeffler** [comment]
Just a heads up, @mitsuhiko + @max-sixty - I have good reason to believe that these changes broke one of our tests. It seems like this version of the `assert_yaml_snapshot` macro takes ownership of the value, while previous versions only borrowed it. I can reproduce the test failures observed in https://github.com/rojo-rbx/rbx-dom/pull/398 on insta 1.36.0 (the version in which the changes in this PR shipped), but not on insta 1.35.1.

**Comment by max-sixty** [comment]
I'll take a look, thanks @kennethloeffler !

**Comment by max-sixty** [comment]
This is fixed in #453, apologies for the temporary break

**Comment by kennethloeffler** [comment]
It's perfectly okay - it's easy to see how this flew under the radar 😅

**Comment by mitsuhiko** [comment]
Sorry for that. I pushed out a fix!

**Pull request merged**